README.md
# Welcome to Microclimate!
Microclimate is a development environment that equips you to create and modernize applications in one seamless experience. Every Microclimate project is set to react to changes immediately, regardless of language, Docker, or your IDE. From writing and testing code locally to building and deployment with a pipeline, Microclimate is ready for action.

If you're new to microservices, Microclimate is a great place to start learning. Our product automates tasks, such as creating the development environment. It also downloads the runtime and build tools that you need to begin working. You don't need to know everything about every area of microservices; you can focus on the applications of your own.

Microclimate applications are fully contained inside Docker containers throughout the development process. Lightweight containers are easily reproducible to match your production environment. Plus, a containerized environment looks the same when you deploy it to the cloud even if you develop and test locally.

With a Kubernetes deployment of Microclimate, you can access a seamless and configurable end-to-end environment from development to deployment. You also have cluster computing advantages, including cluster storage for local or non-committed code and built-in logging.

Microclimate includes the powerful Theia IDE. We’ve extended its capabilities by putting language servers in the image as well as our own XML Language Support. But if you don't want to use a web-based IDE, we've also made your project ready for rapid development with any popular IDE, including Eclipse, Visual Studio Code, and Atom. Point your IDE at the project, and it’s ready to go!

But automation isn't the only way Microclimate helps you save time. Our product is designed to maximize the efficiency of the development inner loop: Code. Build. Deploy. Run. If one step is slow, it delays the rest. But with Microclimate, each step takes only a bare minimum of time thanks to Docker and Kubernetes strategies applied to our supported runtimes. Round-tripping through edit, build, and run allows real-time performance insights, regardless of what development phase you’re in.

Besides helping you save time, Microclimate also offers choices. You can download and start with a free local installation, or you can install it on IBM Cloud Private. Create and import projects in a variety of development languages, including Java, Node.js, and Swift, or you can add your own by importing a template.

For more information, see the [Microclimate website](https://microclimate-dev2ops.github.io/).


_news/news.md
---
layout: news
title: News
description: Latest news
keywords: GA, release, version, history
duration: 20 seconds
permalink: news
order: 19
linkname: Microclimate deprecated
parent: Nov 2019
---

# Microclimate deprecated

*November 1, 2019*

As was announced last year, we have made a significant pivot on our development tools strategy. We learned some lessons developing and delivering Microclimate, and while some of the key value propositions were on point, we felt we missed on how we enabled those solutions for developers. As a result, we moved away from the idea of an all encompassing web-like IDE, and enabled the same key functions through extensions into industry standard IDEs via Eclipse Codewind. Please visit our new home at <a href="https://codewind.dev/">codewind.dev</a> to see what new and exciting content we have delivered. Through Eclipse Codewind we will continue to enable developers with features that will help ease their journey to developing cloud-native applications. Below is the official deprecation announcement for Microclimate:

**Microclimate has officially been deprecated. The evolution of Microclimate will now be delivered via the open source project Eclipse Codewind. In addition to being it's own standalone open source project, Codewind is also integrated with other key open source technologies that make up the commercial product IBM Cloud Paks for Applications. IBM Cloud Pak for Applications is the next generation application development platform that helps developers design, develop, deploy and manage their cloud-native applications with speed and control.**

**On December 1st, 2019 the helm chart for Microclimate will no longer be supported and will be removed from IBM's public helm repository on github.com, which will in turn result in the chart no longer being displayed in the IBM Cloud Private catalog. This will not impact existing deployments of the helm chart.**
 



src/scripts/main.js
$(document).foundation();

$(document).keydown(function(e) {
	if (e.keyCode == 27) {
		if ($('.videoContainer').length) {
			$('.videoContainer').fadeOut(500, function() { $(this).remove(); });
		}
	}
});

function trackVideoPlay(file) {
	window.bluemixAnalytics.trackEvent("Custom Event",{
	    productTitle: digitalData.page.pageInfo.productTitle,
	    category: digitalData.page.pageInfo.analytics.category,
	    action: "View video",
	    customName1: "source",
	    customValue1: "Landing page",
	    objectType: "Video",
	    object: file
    });
}


function trackDownload(file) {
	window.bluemixAnalytics.trackEvent("Downloaded Hybrid Solution",{
	    productTitle: digitalData.page.pageInfo.productTitle,
	    category: digitalData.page.pageInfo.analytics.category,
	    productVersion: "latest",
	    customName2: "fileName",
	    customName1: "source",
	    customValue1: "Landing page",
	    customValue2: file
    });
}

$(document).ready(function(){
    // $.ajax({
    //     url: 'https://api.github.ibm.com/repos/dev-ex/microclimate/tags?access_token=f5fb332a281a15d8165db037b318ac837eff96a2', // The URL to the API. You can get this by clicking on "Show CURL example" from an API profile
    //     type: 'GET', // The HTTP Method
    //     data: {}, // Additional parameters here
    //     dataType: 'json',
    //     success: function(data) { console.log(JSON.stringify(data)); },
    //     error: function(err) { console.log(err); },
    //     beforeSend: function(xhr) {
    //     // xhr.setRequestHeader("X-Mashape-Authorization", "YOUR-MASHAPE-KEY"); // Enter here your Mashape key
    //     }
    // });

    $(".trackdownload").on("click", function(e) {
        var file=$(this).attr('href');
        trackDownload(file);
    });

    // track embeded video play
    $(".embeded-video").on('play',function(){
        var file =  $(this).children("source").attr("src");
        trackVideoPlay(file);   
    });

    // track intro video play
    $("video").on('play',function(){
        var file =  $(this).attr("src");
        trackVideoPlay(file);   
    });
    
    $(".showVideo").on("click", function(e) {
        e.preventDefault();
        var file = $(this).data("video");
        trackVideoPlay(file);
        
        var htmlTemplate = '<div class="videoContainer"><div class="videoPlayer"><div style="display: block; padding-top:56%; width: 100%;">'
                + '</div><button class="closeBtn">X</button><video class="video-iframe" width="100%" height="100%" autobuffer controls autoplay>'
                + '<source id="mp4" src="'
                + file
                + '" type="video/mp4">'
                + '</video></div><div/>';
        $player = $(htmlTemplate);
        $player.find('.closeBtn').on(
                'click',
                function() {
                    $(this).off().closest(
                            '.videoContainer')
                            .fadeOut(500, function() { $(this).remove(); });
                });
        $player.appendTo('body').addClass(
                'dark-bg').fadeIn();
        return false;

    });
    
    
});

utils/migrateProjects.js
//read project infs - loop
const fs = require('fs')
const path = require('path')
const WORKSPACE = '/microclimate-workspace'
const PROJECT_DIR = path.join(WORKSPACE,'.projects');
const USER = 'admin'
const USER_DIR =  path.join(WORKSPACE, USER);

function main (){
  fs.mkdir(USER_DIR, function(){
    readProjectInfs(function(projectInfs){
      updateAndMoveProjects(projectInfs);
    });
    moveConfigAndLogsDirs();
  });

}

function readProjectInfs(cb) {
  fs.readdir(PROJECT_DIR, function(err, files){
    if (err){
      console.error(`Error reading directory ${PROJECT_DIR}`);
      console.error(err);
    } else {
      let projectInfs = []
      for (file of files){
        if (path.extname(file) === '.inf'){
          fs.readFile(path.join(PROJECT_DIR, file), function(err, data){
            if (err){
              console.error(`Error reading file ${file}`);
              console.error(err);
            }
            projectInfs.push(JSON.parse(data));
            if (projectInfs.length === files.length){
              cb(projectInfs);
            }
          });
        }
      }
    }
  });
}

function updateAndMoveProjects(projectInfs){
  //Create new .projects directoy
  const NEW_PROJECT_DIR = path.join(USER_DIR, '.projects');
  fs.mkdir(NEW_PROJECT_DIR, function(){
    for (inf of projectInfs){
      //Add workspace and location
      inf.workspace = path.join(inf.workspace, USER, '/');
      let locOnDisk = `${inf.locOnDisk}/`;
      inf.location = path.join(inf.workspace, inf.name);
      const NEW_INF_PATH = path.join(NEW_PROJECT_DIR,`${inf.name}.inf`);

      //Write the inf to the new directory
      fs.writeFile(NEW_INF_PATH, JSON.stringify(inf), function(err) {
        if (err){
          console.error(`Error writing file ${NEW_INF_PATH}`);
          console.error(err);
        } else {
          console.log(`File written successfully ${NEW_INF_PATH}`)
          //Move associated project directory
          moveProjectDir(inf.name);
        }
      });
    }
  });
}

function moveProjectDir(projectName){
  fs.rename(path.join(WORKSPACE, projectName), path.join(USER_DIR, projectName), function(err){
    if (err){
      console.error(`Error moving project directory for project ${projectName}`);
      console.error(err);
    } else {
      console.log(`Project directory ${projectName} moved successfully`);
    }

  });
}

function moveConfigAndLogsDirs(){
  // Move .config
  fs.rename(path.join(WORKSPACE, '.config'), path.join(USER_DIR, '.config'), function(err){
    if (err){
      console.error(`Error moving .config directory`);
      console.error(err);
    } else {
      console.log(`.config directory successfully moved`);
    }
  });

  // Move .logs
  fs.rename(path.join(WORKSPACE, '.logs'), path.join(USER_DIR, '.logs'), function(err){
    if (err){
      console.error(`Error moving .logs directory`);
      console.error(err);
    } else {
      console.log(`.logs directory successfully moved`);
    }
  });
}

main();


src/scripts/simple-jekyll-search.js
/*!
  * Simple-Jekyll-Search v1.6.3 (https://github.com/christian-fei/Simple-Jekyll-Search)
  * Copyright 2015-2018, Christian Fei
  * Licensed under the MIT License.
  */

(function(){
/* globals ActiveXObject:false */

'use strict'

var _$JSONLoader_2 = {
  load: load
}

function load (location, callback) {
  var xhr = getXHR()
  xhr.open('GET', location, true)
  xhr.onreadystatechange = createStateChangeListener(xhr, callback)
  xhr.send()
}

function createStateChangeListener (xhr, callback) {
  return function () {
    if (xhr.readyState === 4 && xhr.status === 200) {
      try {
        callback(null, JSON.parse(xhr.responseText))
      } catch (err) {
        callback(err, null)
      }
    }
  }
}

function getXHR () {
  return window.XMLHttpRequest ? new window.XMLHttpRequest() : new ActiveXObject('Microsoft.XMLHTTP')
}

'use strict'

var _$OptionsValidator_3 = function OptionsValidator (params) {
  if (!validateParams(params)) {
    throw new Error('-- OptionsValidator: required options missing')
  }

  if (!(this instanceof OptionsValidator)) {
    return new OptionsValidator(params)
  }

  var requiredOptions = params.required

  this.getRequiredOptions = function () {
    return requiredOptions
  }

  this.validate = function (parameters) {
    var errors = []
    requiredOptions.forEach(function (requiredOptionName) {
      if (typeof parameters[requiredOptionName] === 'undefined') {
        errors.push(requiredOptionName)
      }
    })
    return errors
  }

  function validateParams (params) {
    if (!params) {
      return false
    }
    return typeof params.required !== 'undefined' && params.required instanceof Array
  }
}

'use strict';

function fuzzysearch (needle, haystack) {
  var tlen = haystack.length;
  var qlen = needle.length;
  if (qlen > tlen) {
    return false;
  }
  if (qlen === tlen) {
    return needle === haystack;
  }
  outer: for (var i = 0, j = 0; i < qlen; i++) {
    var nch = needle.charCodeAt(i);
    while (j < tlen) {
      if (haystack.charCodeAt(j++) === nch) {
        continue outer;
      }
    }
    return false;
  }
  return true;
}

var _$fuzzysearch_1 = fuzzysearch;

'use strict'

/* removed: var _$fuzzysearch_1 = require('fuzzysearch') */;

var _$FuzzySearchStrategy_5 = new FuzzySearchStrategy()

function FuzzySearchStrategy () {
  this.matches = function (string, crit) {
    return _$fuzzysearch_1(crit.toLowerCase(), string.toLowerCase())
  }
}

'use strict'

var _$LiteralSearchStrategy_6 = new LiteralSearchStrategy()

function LiteralSearchStrategy () {
  this.matches = function (str, crit) {
    if (typeof str !== 'string') {
      return false
    }
    str = str.trim()
    return str.toLowerCase().indexOf(crit.toLowerCase()) >= 0
  }
}

'use strict'

var _$Repository_4 = {
  put: put,
  clear: clear,
  search: search,
  setOptions: setOptions
}

/* removed: var _$FuzzySearchStrategy_5 = require('./SearchStrategies/FuzzySearchStrategy') */;
/* removed: var _$LiteralSearchStrategy_6 = require('./SearchStrategies/LiteralSearchStrategy') */;

function NoSort () {
  return 0
}

var data = []
var opt = {}

opt.fuzzy = false
opt.limit = 10
opt.searchStrategy = opt.fuzzy ? _$FuzzySearchStrategy_5 : _$LiteralSearchStrategy_6
opt.sort = NoSort

function put (data) {
  if (isObject(data)) {
    return addObject(data)
  }
  if (isArray(data)) {
    return addArray(data)
  }
  return undefined
}
function clear () {
  data.length = 0
  return data
}

function isObject (obj) {
  return Boolean(obj) && Object.prototype.toString.call(obj) === '[object Object]'
}

function isArray (obj) {
  return Boolean(obj) && Object.prototype.toString.call(obj) === '[object Array]'
}

function addObject (_data) {
  data.push(_data)
  return data
}

function addArray (_data) {
  var added = []
  clear()
  for (var i = 0, len = _data.length; i < len; i++) {
    if (isObject(_data[i])) {
      added.push(addObject(_data[i]))
    }
  }
  return added
}

function search (crit) {
  if (!crit) {
    return []
  }
  return findMatches(data, crit, opt.searchStrategy, opt).sort(opt.sort)
}

function setOptions (_opt) {
  opt = _opt || {}

  opt.fuzzy = _opt.fuzzy || false
  opt.limit = _opt.limit || 10
  opt.searchStrategy = _opt.fuzzy ? _$FuzzySearchStrategy_5 : _$LiteralSearchStrategy_6
  opt.sort = _opt.sort || NoSort
}

function findMatches (data, crit, strategy, opt) {
  var matches = []
  for (var i = 0; i < data.length && matches.length < opt.limit; i++) {
    var match = findMatchesInObject(data[i], crit, strategy, opt)
    if (match) {
      matches.push(match)
    }
  }
  return matches
}

function findMatchesInObject (obj, crit, strategy, opt) {
  for (var key in obj) {
    if (!isExcluded(obj[key], opt.exclude) && strategy.matches(obj[key], crit)) {
      return obj
    }
  }
}

function isExcluded (term, excludedTerms) {
  var excluded = false
  excludedTerms = excludedTerms || []
  for (var i = 0, len = excludedTerms.length; i < len; i++) {
    var excludedTerm = excludedTerms[i]
    if (!excluded && new RegExp(term).test(excludedTerm)) {
      excluded = true
    }
  }
  return excluded
}

'use strict'

var _$Templater_7 = {
  compile: compile,
  setOptions: __setOptions_7
}

var options = {}
options.pattern = /\{(.*?)\}/g
options.template = ''
options.middleware = function () {}

function __setOptions_7 (_options) {
  options.pattern = _options.pattern || options.pattern
  options.template = _options.template || options.template
  if (typeof _options.middleware === 'function') {
    options.middleware = _options.middleware
  }
}

function compile (data) {
  return options.template.replace(options.pattern, function (match, prop) {
    var value = options.middleware(prop, data[prop], options.template)
    if (typeof value !== 'undefined') {
      return value
    }
    return data[prop] || match
  })
}

'use strict'

var _$utils_9 = {
  merge: merge,
  isJSON: isJSON
}

function merge (defaultParams, mergeParams) {
  var mergedOptions = {}
  for (var option in defaultParams) {
    if (Object.prototype.hasOwnProperty.call(defaultParams, option)) {
      mergedOptions[option] = defaultParams[option]
      if (typeof mergeParams[option] !== 'undefined') {
        mergedOptions[option] = mergeParams[option]
      }
    }
  }
  return mergedOptions
}

function isJSON (json) {
  try {
    if (json instanceof Object && JSON.parse(JSON.stringify(json))) {
      return true
    }
    return false
  } catch (err) {
    return false
  }
}

var _$src_8 = {};
(function (window) {
  'use strict'

  var options = {
    searchInput: null,
    resultsContainer: null,
    json: [],
    success: Function.prototype,
    searchResultTemplate: '<li><a href="{url}" title="{description}">{title}</a></li>',
    templateMiddleware: Function.prototype,
    sortMiddleware: function () {
      return 0
    },
    noResultsText: 'No results found',
    limit: 10,
    fuzzy: false,
    exclude: []
  }

  var requiredOptions = ['searchInput', 'resultsContainer', 'json']

  /* removed: var _$Templater_7 = require('./Templater') */;
  /* removed: var _$Repository_4 = require('./Repository') */;
  /* removed: var _$JSONLoader_2 = require('./JSONLoader') */;
  var optionsValidator = _$OptionsValidator_3({
    required: requiredOptions
  })
  /* removed: var _$utils_9 = require('./utils') */;

  /*
    Public API
  */
  window.SimpleJekyllSearch = function (_options) {
    var errors = optionsValidator.validate(_options)
    if (errors.length > 0) {
      throwError('You must specify the following required options: ' + requiredOptions)
    }

    options = _$utils_9.merge(options, _options)

    _$Templater_7.setOptions({
      template: options.searchResultTemplate,
      middleware: options.templateMiddleware
    })

    _$Repository_4.setOptions({
      fuzzy: options.fuzzy,
      limit: options.limit,
      sort: options.sortMiddleware
    })

    if (_$utils_9.isJSON(options.json)) {
      initWithJSON(options.json)
    } else {
      initWithURL(options.json)
    }

    return {
      search: search
    }
  }

  // For backwards compatibility
  window.SimpleJekyllSearch.init = window.SimpleJekyllSearch

  if (typeof window.SimpleJekyllSearchInit === 'function') {
    window.SimpleJekyllSearchInit.call(this, window.SimpleJekyllSearch)
  }

  function initWithJSON (json) {
    options.success(json)
    _$Repository_4.put(json)
    registerInput()
  }

  function initWithURL (url) {
    _$JSONLoader_2.load(url, function (err, json) {
      if (err) {
        throwError('failed to get JSON (' + url + ')')
      }
      initWithJSON(json)
    })
  }

  function emptyResultsContainer () {
    options.resultsContainer.innerHTML = ''
  }

  function appendToResultsContainer (text) {
    options.resultsContainer.innerHTML += text
  }

  function registerInput () {
    options.searchInput.addEventListener('keyup', function (e) {
      if (isWhitelistedKey(e.which)) {
        emptyResultsContainer()
        search(e.target.value)
      }
    })
  }

  function search (query) {
    if (isValidQuery(query)) {
      emptyResultsContainer()
      render(_$Repository_4.search(query))
    }
  }

  function render (results) {
    var len = results.length
    if (len === 0) {
      return appendToResultsContainer(options.noResultsText)
    }
    for (var i = 0; i < len; i++) {
      appendToResultsContainer(_$Templater_7.compile(results[i]))
    }
  }

  function isValidQuery (query) {
    return query && query.length > 0
  }

  function isWhitelistedKey (key) {
    return [13, 16, 20, 37, 38, 39, 40, 91].indexOf(key) === -1
  }

  function throwError (message) {
    throw new Error('SimpleJekyllSearch --- ' + message)
  }
})(window)

}());


_site/js/scripts.js
$(document).keydown(function(e) {
	if (e.keyCode == 27) {
		if ($('.videoContainer').length) {
			$('.videoContainer').fadeOut(500, function() { $(this).remove(); });
		}
	}
});

function trackVideoPlay(file) {
	window.bluemixAnalytics.trackEvent("Custom Event",{
	    productTitle: digitalData.page.pageInfo.productTitle,
	    category: digitalData.page.pageInfo.analytics.category,
	    action: "View video",
	    customName1: "source",
	    customValue1: "Landing page",
	    objectType: "Video",
	    object: file
    });
}


function trackDownload(file) {
	window.bluemixAnalytics.trackEvent("Downloaded Hybrid Solution",{
	    productTitle: digitalData.page.pageInfo.productTitle,
	    category: digitalData.page.pageInfo.analytics.category,
	    productVersion: "latest",
	    customName2: "fileName",
	    customName1: "source",
	    customValue1: "Landing page",
	    customValue2: file
    });
}




$(document)
		.ready(
				function() {
					
					$(".trackdownload").on("click", function(e) {
						var file=$(this).attr('href');
						
						trackDownload(file);
					});
					
					$(".navbar").on("show.bs.collapse", function(e) {
						  console.log("open")
							$(".navbar").css("background", "black");
						  $(".navbar-collapse").css("background", "black");
						});

						$(".navbar").on("hide.bs.collapse", function(e) {
						  console.log("close")
							$(".navbar").css("background", "transparent");
						    $(".navbar-collapse").css("background", "transparent");
						});
					// track embeded video play
					$(".embeded-video").on('play',function(){
						var file =  $(this).children("source").attr("src");
						trackVideoPlay(file);
						
						
					});
					$(".showVideo")
							.on(
									"click",
									function(e) {
										e.preventDefault();
										var file = $(this).data("video");
										trackVideoPlay(file);
									    
										var htmlTemplate = '<div class="videoContainer"><div class="videoPlayer"><div style="display: block; padding-top:56%; width: 100%;">'
												+ '</div><button class="closeBtn">X</button><video class="video-iframe" width="100%" height="100%" autobuffer controls autoplay>'
												+ '<source id="mp4" src="'
												+ file
												+ '" type="video/mp4">'
												+ '</video></div><div/>';
										$player = $(htmlTemplate);
										$player.find('.closeBtn').on(
												'click',
												function() {
													$(this).off().closest(
															'.videoContainer')
															.fadeOut(500, function() { $(this).remove(); });
												});
										$player.appendTo('body').addClass(
												'dark-bg').fadeIn();
										return false;

									});

					
				});

	
